includes/class-wc-data-structurer.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+/**
+ * Structured data's handler and generator using JSON-LD format.
+ *
+ * @class 		WC_Data_Structurer
+ * @version		2.7.0
+ * @package		WooCommerce/Classes
+ * @category	Class
+ * @author 		Clement Cazaud
+ */
+class WC_Data_Structurer {
+  
+  /**
+   * @var array Partially formatted structured data
+   */
+  private $structured_data;
+
+  /**
+   * Checks if the passed $json variable is an array and stores it into $this->structured_data...
+   *
+   * @param array $json Partially formatted JSON-LD
+   * @return false If the param $json is not an array
+   */
+  private function set_structured_data( $json ) {
+    if ( ! is_array( $json ) ) {
+      return false;
+    }
+    $this->structured_data[] = $json;
+  }
+  
+  /**
+   * Formats and returns the structured data...
+   *
+   * @return mixed bool|array If structured data is set, returns the fully formatted and encoded JSON-LD, otherwise returns false
+   */
+  private function get_structured_data() {
+    if ( ! $this->structured_data ) {
+      return false;
+    }
+    $product_count = 0;
+
+    foreach ( $this->structured_data as $value ) {
+      $type = isset( $value['@type'] ) ? $value['@type'] : false;
+        
+      if ( 'Product' === $type || 'SoftwareApplication' === $type || 'MusicAlbum' === $type ) {
+        $products[] = $value;
+        $product_count ++;
+      }
+      elseif ( 'Review' === $type ) {
+        $reviews[] = $value;
+      }
+    }
+    if ( $product_count === 1 ) {
+      $structured_data = isset( $reviews ) ? $products[0] + array( 'review' => $reviews ) : $products[0];
+    }
+    elseif ( $product_count > 1 ) {
+      $structured_data = array( '@graph' => $products );
+    }
+    if ( ! isset( $structured_data ) ) {
+      return false;
+    }
+    $context['@context'] = 'http://schema.org/';
+    
+    return wp_json_encode( $context + $structured_data );
+  }
+
+  /**
+   * Contructor
+   */
+  public function __construct() {
+    add_action( 'woocommerce_before_shop_loop_item', array( $this, 'init_product_category_structured_data' ) );
+    add_action( 'woocommerce_single_product_summary', array( $this, 'init_product_structured_data' ) );
+    add_action( 'woocommerce_review_meta', array( $this, 'init_product_review_structured_data' ) );
+    add_action( 'wp_footer', array( $this, 'enqueue_structured_data' ) );
```

Comment:
@mikejolley 

If I get you right, in the actual implementation, this is not possible because there can be multiple of these elements being displayed on the same page. Eg: On single product page, the instance generates 'Product' type structured data and potentially multiple 'Review' items. Which all of them have then to be wrapped into a single array for the output...

But maybe I'm missing something ?